py/dml/dmlparse.py

```diff
@@ -2350,23 +2371,27 @@ def hashselect(t):
 @prod
 def select(t):
     'statement_except_hashif : hashselect ident IN LPAREN expression RPAREN WHERE LPAREN expression RPAREN statement hashelse statement'
+    ident_enforce_not_discard_ref(site(t, 2), t[2])
     t[0] = ast.select(site(t), t[2], t[5], t[9], t[11], t[13])
 
 @prod_dml12
 def foreach(t):
     'statement_except_hashif : FOREACH ident IN LPAREN expression RPAREN statement'
+    ident_enforce_not_discard_ref(site(t, 2), t[2])
     t[0] = ast.foreach_dml12(site(t), t[2], t[5], t[7])
     if logging.show_porting:
         report(PHASH(site(t)))
 
 @prod_dml14
 def foreach(t):
     'statement_except_hashif : FOREACH ident IN LPAREN expression RPAREN statement'
+    ident_enforce_not_discard_ref(site(t, 2), t[2])
     t[0] = ast.foreach(site(t), t[2], t[5], t[7])
 
 @prod_dml14
 def hashforeach(t):
     'statement_except_hashif : HASHFOREACH ident IN LPAREN expression RPAREN statement'
+    ident_enforce_not_discard_ref(site(t, 2), t[2])
```

Comment:
Lots of these calls. Perhaps easier to make a single call in the `ident` production rule? And create a separate `ident_or_underscore` production rule, which would be like `ident` but without that call?

Reply:
Nope. I tried, but ran into reduce/reduce conflicts.

Reply:
Oh, but there should be some middle path: There should be no reduce/reduce after `HASHFOREACH`, so there you can use `ident`, and in other cases (the start token of `statement`, I suppose) there will be reduce/reduce so for those can do `ident_or_underscore` with a manual `ident_enforce_not_discard_ref`. Could make sense if it's not a vast majority of all cases?

Reply:
Such ad-hoc approaches would be very fragile to parser changes (i.e. a change could force rewriting an existing `ident` to `ident_or_underscore`) not to mention would introduce an inconsistency (_either_ `ident_or_underscore` is used or a manual `ident_enforce_not_discard_ref` is.) This would only serve to make the parser even more difficult to understand. So I feel it'd be worse.

I agree that the `ident_enforce_not_discard_ref` is sub-par, and would welcome any superior approach _which is at least as general and consistently applicable._ `ident_decl` (i.e. `ident` and not `ident_or_underscore`) would be preferrable, if LALR didn't ruin our day.

Reply:
I would argue the opposite wrt fragility: your current approach requires that any new rule that uses `ident` remembers to call `ident_enforce_not_discard_ref`, if not the deprecation will silently be incomplete.

But we probably just misunderstand each other. More efficient if I try to code up what I mean, perhaps I will understand better when I see the R/R conflicts myself.

Reply:
I made an attempt, now pushed to this PR.